segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+                SegmentMetadata sourceSegmentMetadata = (SegmentMetadata) txn.get(sourceSegment);
+                checkSegmentExists(sourceSegment, sourceSegmentMetadata);
+
+                sourceSegmentMetadata.checkInvariants();
+
+                // This is a critical assumption at this point which should not be broken,
+                Preconditions.checkState(!targetSegmentMetadata.isStorageSystemSegment(), "Storage system segments cannot be concatenated.");
+                Preconditions.checkState(!sourceSegmentMetadata.isStorageSystemSegment(), "Storage system segments cannot be concatenated.");
+
+                if (!sourceSegmentMetadata.isSealed()) {
+                    throw new IllegalStateException();
+                }
+
+                if (targetSegmentMetadata.getOwnerEpoch() > this.epoch) {
+                    throw new StorageNotPrimaryException(targetSegmentMetadata.getName());
+                }
+
+                if (sourceSegmentMetadata.getStartOffset() != 0) {
+                    throw new StreamSegmentTruncatedException(sourceSegment, sourceSegmentMetadata.getLength(), 0);
+                }
+
+                if (offset != targetSegmentMetadata.getLength()) {
+                    throw new BadOffsetException(targetHandle.getSegmentName(), targetSegmentMetadata.getLength(), offset);
+                }
+
+                // Update list of chunks by appending sources list of chunks.
+                ChunkMetadata targetLastChunk = (ChunkMetadata) txn.get(targetSegmentMetadata.getLastChunk());
+                ChunkMetadata sourceFirstChunk = (ChunkMetadata) txn.get(sourceSegmentMetadata.getFirstChunk());
+
+                if (targetLastChunk != null) {
+                    targetLastChunk.setNextChunk(sourceFirstChunk.getName());
+                    txn.update(targetLastChunk);
+                } else {
+                    if (sourceFirstChunk != null) {
+                        targetSegmentMetadata.setFirstChunk(sourceFirstChunk.getName());
+                        txn.update(sourceFirstChunk);
+                    }
+                }
+
+                // Update segments's last chunk to point to the sources last segment.
+                targetSegmentMetadata.setLastChunk(sourceSegmentMetadata.getLastChunk());
+
+                // Update the length of segment.
+                targetSegmentMetadata.setLastChunkStartOffset(targetSegmentMetadata.getLength() + sourceSegmentMetadata.getLastChunkStartOffset());
+                targetSegmentMetadata.setLength(targetSegmentMetadata.getLength() + sourceSegmentMetadata.getLength() - sourceSegmentMetadata.getStartOffset());
+
+                targetSegmentMetadata.setChunkCount(targetSegmentMetadata.getChunkCount() + sourceSegmentMetadata.getChunkCount());
+
+                txn.update(targetSegmentMetadata);
+                txn.delete(sourceSegment);
+
+                // Finally defrag immediately.
+                if (shouldDefrag() && null != targetLastChunk) {
```

Comment:
Looking at `shouldDefrag`, I'm not sure what the logic is that leads defragmentation. Specifically, the javadoc for the defrag call says that it expects a list of chunks, so I'm not clear on what the condition is.

Reply:
List as in range between - String startChunkName, String lastChunkName.

Reply:
Currently it is a simple condition - if underlying store supports concat or supports appends to same chunk. However in future it could be more complex condition.